Add GE-Spot Support to EV Smart Charging

enoch85 opened the pull request:
This PR adds support for the GE-Spot integration as a price source for EV Smart Charging, alongside the existing Nordpool and Energi Data Service integrations.

## What is GE-Spot?

[GE-Spot (Global Electricity Spot Prices) ](https://github.com/enoch85/ge-spot/blob/main/README.md)is a Home Assistant integration that provides standardized electricity price sensors for multiple regions and markets. It supports various price sources including Nordpool, ENTSO-E, Energi Data Service, and more, with robust fallback mechanisms and timezone handling.

## Changes Made

1. **Added GE-Spot as a supported platform**
   - Added `PLATFORM_GESPOT = "ge_spot"` constant in `const.py`
   - Updated imports in relevant files to include this constant

2. **Updated the config flow**
   - Modified the config flow to check for GE-Spot sensors if both Nordpool and Energi Data Service sensors are not found
   - This provides a seamless fallback mechanism for users with GE-Spot installed

3. **Added helper functions**
   - Added `find_gespot_sensor` function to find GE-Spot sensors in the entity registry
   - This follows the same pattern as the existing `find_nordpool_sensor` and `find_energidataservice_sensor` functions

4. **Updated the coordinator**
   - Added GE-Spot case in the `Raw.__init__` method to handle GE-Spot price data format
   - Since GE-Spot has the same data format as Nordpool (with "value" and "start" fields), we're handling it the same way
   - Updated the coordinator to pass the price platform parameter when creating Raw objects

5. **Enhanced the validator**
   - Updated the `Validator.is_price_state` method to handle GE-Spot price data format
   - Added special handling for GE-Spot to support different data formats based on its configured primary source
   - Added support for GE-Spot's fallback mechanism, allowing validation to succeed even when using fallback sources
   - This ensures that GE-Spot price entities are properly validated regardless of their configuration

6. **Updated the README**
   - Added GE-Spot to the list of supported integrations in the README.md file

## Tests Added

Comprehensive tests have been added to ensure the functionality works correctly:

1. **Test data for GE-Spot** in `tests/price.py`
   - Added test data for GE-Spot that follows the same format as Nordpool

2. **Tests for the Raw class** with GE-Spot data format:
   - Added `test_raw_gespot` in `tests/helpers/test_coordinator.py`
   - Added `test_raw_gespot_simple.py` with detailed tests for the Raw class with GE-Spot data
   - Tests verify that the Raw class correctly processes GE-Spot data, validates it, and provides the expected methods

3. **Tests for the FindEntity.find_gespot_sensor function**:
   - Added tests in `tests/helpers/test_config_flow.py`
   - Added `test_find_gespot_simple.py` with detailed tests for finding GE-Spot sensors
   - Tests verify that the function correctly finds GE-Spot sensors in the entity registry and handles edge cases

4. **MockPriceEntityGESpot class** for testing in `tests/helpers/helpers.py`
   - Added a mock class for GE-Spot entities to use in tests

5. **Additional test files**:
   - Added `test_raw_gespot.py` for testing the Raw class with GE-Spot data
   - Added `test_find_gespot.py` for testing the FindEntity.find_gespot_sensor function
   - Added `test_validator_gespot.py` for testing the Validator.is_price_state method with GE-Spot data
   - Added `test_mock_gespot.py` for testing the MockPriceEntityGESpot class

## Benefits

1. **Expanded Geographical Coverage**: GE-Spot supports more regions than Nordpool or Energi Data Service alone
2. **Enhanced Reliability**: GE-Spot's fallback mechanism ensures price data is available even if one source fails
3. **Consistent Interface**: Users get the same experience regardless of their region
4. **Future-Proof**: As GE-Spot adds more price sources, EV Smart Charging automatically benefits
5. **Simplified Configuration**: Users in regions not covered by Nordpool can now use EV Smart Charging
6. **Robust Validation**: The enhanced validation ensures compatibility with different GE-Spot configurations

## Implementation Details

The implementation follows the same pattern as PR #108 which added Energi Data Service support, but with additional enhancements to handle GE-Spot's more flexible configuration options. While GE-Spot's default data format is the same as Nordpool (with "value" and "start" fields), the validation has been improved to handle different data formats and fallback mechanisms, making the integration more robust.

## Testing

The changes have been thoroughly tested with the added test files. All tests pass, ensuring that GE-Spot integration works correctly with EV Smart Charging. The tests verify that:

1. The Raw class correctly handles GE-Spot data format
2. The FindEntity.find_gespot_sensor function correctly finds GE-Spot sensors
3. The Validator.is_price_state method correctly validates GE-Spot state with:
   - Different data formats (Nordpool, Energi Data Service, ENTSOE)
   - Fallback source configurations
   - Various edge cases

enoch85 commented:
@jonasbkarlsson I hope you like it :)

enoch85 commented:
Tests are failing, will fix

jonasbkarlsson commented:
Thanks! Will have a look. But it might take some time. I have a long-overdue solar charging function I need to get ready first.

enoch85 commented:
I'm now done with the refactor, and will let it cool down for a while: https://github.com/enoch85/ge-spot/pull/4

Feel free to discuss this implementation further. The benefits listed are real, and my goal is to work on adding even more APIs in the future.

jonasbkarlsson commented:
Yes, sorry, I have had limited time lately, and solar charging function I mentioned before is a complication.I should probably simplify that one so it will be possible to release it.

enoch85 commented:
@jonasbkarlsson OK, so I added support for 15 min intervals. An upcoming change in the GE-Spot integration.

enoch85 commented:
In the last commit, I made sure all the test were passing and added logic for 15 min and 1 hour intervals in raw.py.

Additionally, I fixed some pre-existing bugs in the core code:
- Fixed interval detection logic to properly handle hourly data conversion to 15-minute intervals
- Corrected test data inconsistencies (PRICE_20220930_ENERGIDATASERVICE, PRICE_20221001)
- Removed dead code (30-minute interval support)
- Improved type hints to be more accurate

Other than that I made sure that the GE Spot price sensor actually works in the EV Smart Charging integration by doing real life tests with a release based on this PR. I can now say 100% that this is working as expected.

Here are the test results:

``` 
tests/helpers/test_price_adaptor.py::test_inintiate ✓                                                                                       3% ▍         
 tests/helpers/test_price_adaptor.py::test_is_price_state ✓                                                                                  6% ▋         
 tests/helpers/test_price_adaptor.py::test_get_raw_today_local ✓                                                                            10% █         
 tests/helpers/test_price_adaptor.py::test_get_raw_tomorrow_local ✓                                                                         13% █▍        
 tests/helpers/test_price_adaptor.py::test_get_current_price ✓                                                                              16% █▋        
 tests/helpers/test_price_adaptor.py::test_validate_price_entity ✓                                                                          19% █▉        
 tests/helpers/test_price_adaptor.py::test_validate_price_entity_entsoe ✓                                                                   23% ██▍       
 tests/helpers/test_coordinator.py::test_raw ✓                                                                                              26% ██▋       
 tests/helpers/test_coordinator.py::test_raw_energidataservice ✓                                                                            29% ██▉       
 tests/helpers/test_coordinator.py::test_raw_entsoe ✓                                                                                       32% ███▎      
 tests/helpers/test_coordinator.py::test_raw_tge ✓                                                                                          35% ███▋      
 tests/helpers/test_coordinator.py::test_get_lowest_quarters_non_continuous ✓                                                               39% ███▉      
 tests/helpers/test_coordinator.py::test_get_lowest_quarters_non_continuous_15min ✓                                                         42% ████▎     
 tests/helpers/test_coordinator.py::test_get_lowest_quarters_continuous ✓                                                                   45% ████▌     
 tests/helpers/test_coordinator.py::test_get_lowest_quarters_continuous_15min ✓                                                             48% ████▉     
 tests/helpers/test_coordinator.py::test_get_charging_original ✓                                                                            52% █████▎    
 tests/helpers/test_coordinator.py::test_get_charging_update ✓                                                                              55% █████▌    
 tests/helpers/test_coordinator.py::test_get_charging_quarters ✓                                                                            58% █████▊    
 tests/helpers/test_coordinator.py::test_get_charging_value ✓                                                                               61% ██████▎   
 tests/helpers/test_coordinator.py::test_scheduler ✓                                                                                        65% ██████▌   
 tests/helpers/test_coordinator.py::test_get_empty_schedule ✓                                                                               68% ██████▊   
 tests/helpers/test_coordinator.py::test_get_ready_quarter_utc ✓                                                                            71% ███████▏  
 tests/helpers/test_coordinator.py::test_get_start_quarter_utc ✓                                                                            74% ███████▌  
 tests/test_gespot.py::TestGESpotIntegration.test_find_gespot_sensor ✓                                                                      77% ███████▊  
 tests/test_gespot.py::TestGESpotIntegration.test_validator_gespot ✓                                                                        81% ████████▏ 
 tests/test_gespot.py::TestGESpotIntegration.test_96_interval_data ✓                                                                        84% ████████▍ 
 tests/test_gespot.py::TestGESpotIntegration.test_raw_price_handler_96_intervals ✓                                                          87% ████████▊ 
 tests/test_gespot_dst.py::TestGESpotDST.test_normal_day_96_intervals ✓                                                                     90% █████████ 
 tests/test_gespot_dst.py::TestGESpotDST.test_dst_spring_92_intervals ✓                                                                     94% █████████▍
 tests/test_gespot_dst.py::TestGESpotDST.test_dst_fall_100_intervals ✓                                                                      97% █████████▋
 tests/test_gespot_dst.py::TestGESpotDST.test_dst_edge_case_validation ✓                                                                   100% ██████████

Results (1.82s):
      31 passed
      ```

enoch85 commented:
@jonasbkarlsson This is now 100% done. Please release it as a beta so that we have good test coverage before merging to prod.  Even if I tested this myself, it's better to be safe than sorry. :)

enoch85 commented:
@jonasbkarlsson You can test with this release. It's running in my PROD now. :)

https://github.com/enoch85/ev_smart_charging_temp/releases/tag/v0.4.0-rc1

enoch85 commented:
OK, so this is now running in my production,  and it works! Pic or it didn't happen:

<img width="283" height="260" alt="image" src="https://github.com/user-attachments/assets/5cea5518-501d-4159-9979-8bb47481b425" />

<img width="526" height="330" alt="image" src="https://github.com/user-attachments/assets/41522f47-c256-41a8-b8d5-af5c2effec21" />

jonasbkarlsson commented:
I looked through the PR (except test test parts), and I have number of comments. I need to look more, but here are two first comments.

The changes in `ev_smart_charging/config_flow.py` are not needed, since you have updated `FindEntity.find_price_sensor()`. 

In `Raw.__init__()` there is a new part added (lines 146-178) that handles a special case. However, the orginal code handles that case as well. So as a first step I would suggest to remove lines 146-178. And then the case with `platform == None` needs to be handled, e.g., by `platform: str | PriceFormat | None = PLATFORM_NORDPOOL`

enoch85 commented:
@jonasbkarlsson 

OK, thanks! Fixed your review comments.

```
======================== 122 passed in 72.74s (0:01:12) ========================
```

enoch85 commented:
OK, I'm done. https://github.com/enoch85/ge-spot/pull/22

We now use this as our main sensors in GE-Spot. Thanks for pushing me in the right direction. :+1:

enoch85 commented:
@jonasbkarlsson 

Results (50.49s):
     122 passed

enoch85 commented:
> Great! I now looked more carefully at the test code. I noticed that you had created a new PRICE_20221001_THIRTEEN, and used that instead of PRICE_THIRTEEN_LIST in some places, but not everywhere.
> 
> What is the purpose of PRICE_20221001_THIRTEEN, compared with the PRICE_THIRTEEN_LIST? (I'm not suggesting that it's not needed. I'm just asking because I have not taking the time to understand...).

@jonasbkarlsson Honestly I don't remember myself. That day I was super tired, and I messed up all the tests, so there might have been some confusion regarding that. 

I've been working on several other projects as well since, so I'm quite dizzy with the timeline and what is done where and so on (sleept 5 hours last 2 days).

But I do remember there was some inconsistency that I was trying to solve. Now when looking again the only thing I changed was the actual list name, so not sure. Sorry! Feel free to leave it, or use it - all tests are passing, that's what I care about. :)

EDIT: 
### 13-hour subset used in adaptor tests (mimics shortened tomorrow list)
### PRICE_20221001_THIRTEEN = PRICE_20221001[:13]

jonasbkarlsson commented:
Included in [v2.5.0](https://github.com/jonasbkarlsson/ev_smart_charging/releases/tag/v2.5.0). Thanks for the contribution, and with your patience with my initial slow response.
